contracts/upgradeable_contracts/BasicOmnibridge.sol

```diff
@@ -297,17 +368,78 @@ abstract contract BasicOmnibridge is
+     * @param _balanceChange amount of balance to subtract from the mediator balance.
      */
     function _releaseTokens(
+        bool _isNative,
         address _token,
         address _recipient,
-        uint256 _value
+        uint256 _value,
+        uint256 _balanceChange
     ) internal virtual {
-        IERC677(_token).safeTransfer(_recipient, _value);
-        _setMediatorBalance(_token, mediatorBalance(_token).sub(_value));
+        if (_isNative) {
+            IERC677(_token).safeTransfer(_recipient, _value);
+            _setMediatorBalance(_token, mediatorBalance(_token).sub(_balanceChange));
+        } else {
+            _getMinterFor(_token).mint(_recipient, _value);
+        }
+    }
+
+    /**
+     * Internal function getting address of the bridged token. Deploys new token is necessary.
+     * @param _token address of the token contract on the other side of the bridge.
+     * @param _name name of the native token, name suffix will be appended, if empty, symbol will be used instead.
+     * @param _symbol symbol of the bridged token, if empty, name will be used instead.
+     * @param _decimals decimals of the bridge foreign token.
+     */
+    function _getBridgedTokenOrDeploy(
+        address _token,
+        string calldata _name,
+        string calldata _symbol,
+        uint8 _decimals
+    ) internal returns (address) {
+        address bridgedToken = bridgedTokenAddress(_token);
+        if (bridgedToken == address(0)) {
+            string memory name = _name;
+            string memory symbol = _symbol;
+            require(bytes(name).length > 0 || bytes(symbol).length > 0);
+            if (bytes(name).length == 0) {
+                name = symbol;
+            } else if (bytes(symbol).length == 0) {
+                symbol = name;
+            }
+            name = _transformName(name);
+            bridgedToken = tokenFactory().deploy(name, symbol, _decimals, bridgeContract().sourceChainId());
+            _setTokenAddressPair(_token, bridgedToken);
+            _initializeTokenBridgeLimits(bridgedToken, _decimals);
+        } else if (!isTokenRegistered(bridgedToken)) {
+            require(IERC20Metadata(bridgedToken).decimals() == _decimals);
+            _initializeTokenBridgeLimits(bridgedToken, _decimals);
+        }
+        return bridgedToken;
+    }
+
+    /**
+     * Notifies receiving contract about the completed bridging operation.
+     * @param _recipient address of the tokens receiver.
+     * @param _token address of the bridged token.
+     * @param _value amount of tokens transferred.
+     * @param _data additional data passed to the callback.
+     */
+    function _receiverCallback(
+        address _recipient,
+        address _token,
+        uint256 _value,
+        bytes memory _data
+    ) internal {
+        if (Address.isContract(_recipient)) {
+            bytes memory data = abi.encodeWithSelector(IERC20Receiver.onTokenTransfer.selector, _token, _value, _data);
```

Comment:
Since `onTokenTransfer` is from the ERC677 standard it is undesirable to change meaning of its parameters. Should we introduce a new method instead (e.g. `onTokenBridged`)?

Reply:
Agree, I have renamed it to `onTokenBridged`.